src/datasource.ts

```diff
@@ -356,28 +356,37 @@ export function getRange(range: TimeRange): Tuple<number> {
   return [timeFrom, timeTo];
 }
 
-function showTooMuchDatapointsWarningIfNeeded(
-  responses: SuccessResponse<ResponseMetadata, DataQueryRequestResponse>[]
-) {
-  responses.forEach(
-    ({
-      result: {
-        data: { items },
-        config: {
-          data: { limit },
-        },
-      },
-      metadata: {
-        target: { refId },
+function showWarnings(responses: SuccessResponse<ResponseMetadata, DataQueryRequestResponse>[]) {
+  responses.forEach(({ result, metadata }) => {
+    const {
+      data: { items },
+      config: {
+        data: { limit },
       },
-    }) => {
-      const hasMorePoints = items.some(({ datapoints }) => datapoints.length >= limit);
-      const warning =
-        '[WARNING] Datapoints limit was reached, so not all datapoints may be shown. Try increasing the granularity, or choose a smaller time range.';
-
-      if (hasMorePoints) {
-        appEvents.emit(datapointsLimitWarningEvent, { refId, warning });
-      }
+    } = result;
+    const {
+      target: { refId },
+    } = metadata;
+    const hasMorePoints = items.some(({ datapoints }) => datapoints.length >= limit);
+    const warning =
+      'Datapoints limit was reached, so not all datapoints may be shown. Try increasing the granularity, or choose a smaller time range.';
+
+    if (hasMorePoints) {
+      appEvents.emit(datapointsWarningEvent, { refId, warning });
     }
-  );
+
+    const calculationErrors = new Set();
+    items.forEach(({ datapoints }) => {
+      (datapoints as [])
+        .map(({ error }) => error)
+        .filter(Boolean)
+        .forEach(error => {
+          calculationErrors.add(error);
+        });
+    });
+
+    calculationErrors.forEach(message => {
+      appEvents.emit(datapointsWarningEvent, { refId, warning: message });
```

Comment:
mb we can send only one event here but with the array of warnings?

Reply:
fixed, but I've sent it as one string
should be ok as well